parser-typechecker/src/Unison/Codebase/Branch.hs

```diff
@@ -586,40 +615,73 @@ modifyAtM path f b = case Path.uncons path of
-
-    -- groups the actions based on the child they apply to
-    childActions :: Map NameSegment [(Path, Branch0 m -> n (Branch0 m))]
-    childActions =
-      List.multimap [ (seg, (rest,f)) | (seg :< rest, f) <- actions ]
-
-    -- alters the children of `b` based on the `childActions` map
-    stepChildren :: Map NameSegment (Branch m) -> n (Map NameSegment (Branch m))
-    stepChildren children0 = foldM g children0 $ Map.toList childActions
+batchUpdates actions =
+  runIdentity . batchUpdatesM actionsIdentity
+  where
+    actionsIdentity :: [(Path, Branch0 m -> Identity (Branch0 m))]
+    actionsIdentity = coerce $ toList actions
+
+data ActionLocation = HereActions | ChildActions
+  deriving Eq
+
+-- | Efficiently perform updates over many locations within a branch.
+-- All updates are performed over the current branch, all causal changes
+-- must be performed in the updates themselves.
+batchUpdatesM ::
+  forall m n f.
+  (Monad m, Monad n, Foldable f) =>
+  f (Path, Branch0 m -> n (Branch0 m)) ->
+  Branch0 m ->
+  n (Branch0 m)
+batchUpdatesM (toList -> actions) curBranch = foldM execActions curBranch (groupActionsByLocation actions)
+  where
+    groupActionsByLocation :: [(Path, b)] -> [(ActionLocation, [(Path, b)])]
+    groupActionsByLocation = List.groupMap \(p, act) -> (pathLocation p, (p, act))
+
+    execActions ::
+      ( Branch0 m ->
+        (ActionLocation, [(Path, Branch0 m -> n (Branch0 m))]) ->
+        n (Branch0 m)
+      )
+    execActions b = \case
+      (HereActions, acts) -> foldM (\b (_, act) -> act b) b acts
+      (ChildActions, acts) -> b & children %%~ stepChildren (groupByNextSegment acts)
+
+    stepChildren ::
+      Map NameSegment [(Path, Branch0 m -> n (Branch0 m))] ->
+      Map NameSegment (Branch m) ->
+      n (Map NameSegment (Branch m))
+    stepChildren childActions children0 =
+      foldM go children0 $ Map.toList childActions
       where
-      g children (seg, actions) = do
         -- Recursively applies the relevant actions to the child branch
-        -- The `findWithDefault` is important - it allows the stepManyAt
-        -- to create new children at paths that don't previously exist.
-        child <- stepM (go actions) (Map.findWithDefault empty seg children0)
-        pure $ updateChildren seg child children
-    in do
-      c2 <- stepChildren (view children b)
-      currentAction (set children c2 b)
+        go ::
+          ( Map NameSegment (Branch m) ->
+            (NameSegment, [(Path, Branch0 m -> n (Branch0 m))]) ->
+            n (Map NameSegment (Branch m))
+          )
+        go children (seg, acts) = do
+          -- 'non empty' creates an empty branch if one is missing,
+          -- and similarly deletes a branch if it is empty after modifications.
+          -- This is important so that branch actions can create/delete branches.
+          children & at seg . non empty . head_ %%~ batchUpdatesM acts
```

Comment:
Note here that we no longer `step` the children, we update the head as-is, it's not necessary to step and would add extra history that would just get squashed down anyways.